src/NuGet.Core/NuGet.CommandLine.XPlat/ListPackage/ListPackageJsonRenderer.cs

```diff
@@ -0,0 +1,400 @@
+
+        protected readonly List<ReportProblem> _problems = new();
+        protected TextWriter _writer;
+
+        private ListPackageJsonRenderer()
+        { }
+
+        public ListPackageJsonRenderer(TextWriter textWriter = null)
+        {
+            _writer = textWriter != null ? textWriter : Console.Out;
+        }
+
+        public void AddProblem(ProblemType problemType, string text)
+        {
+            _problems.Add(new ReportProblem(problemType, string.Empty, text));
+        }
+
+        public IEnumerable<ReportProblem> GetProblems()
+        {
+            return _problems;
+        }
+
+        public void Render(ListPackageReportModel listPackageReportModel)
+        {
+            // Aggregate problems from projects.
+            _problems.AddRange(listPackageReportModel.Projects.Where(p => p.ProjectProblems != null).SelectMany(p => p.ProjectProblems).Where(p => p.ProblemType != ProblemType.Information));
+            var jsonRenderedOutput = WriteJson(new ListPackageOutputContent()
+            {
+                ListPackageArgs = listPackageReportModel.ListPackageArgs,
+                Problems = _problems,
+                Projects = listPackageReportModel.Projects
+            });
+
+            _writer.WriteLine(jsonRenderedOutput);
+        }
+
+        internal string WriteJson(ListPackageOutputContent jsonOutputContent)
+        {
+            ListPackageArgs = jsonOutputContent.ListPackageArgs;
+
+            using (var writer = new StringWriter())
+            {
+                using (var jsonWriter = new JsonTextWriter(writer))
+                {
+                    jsonWriter.Formatting = Formatting.Indented;
+                    _jsonSerializer.Serialize(jsonWriter, jsonOutputContent);
+                }
+
+                return writer.ToString();
+            }
+        }
+
+        private static JsonSerializerSettings GetSerializerSettings()
+        {
+            var settings = new JsonSerializerSettings()
+            {
+                Formatting = Formatting.Indented
+            };
+            settings.Converters.Add(JsonOutputConverter.Default);
+            return settings;
+        }
+
+        private static void WriteProblems(JsonWriter writer, IEnumerable<ReportProblem> reportProblems)
+        {
+            writer.WritePropertyName(ProblemsProperty);
+            writer.WriteStartArray();
+
+            foreach (ReportProblem reportProblem in reportProblems)
+            {
+                writer.WriteStartObject();
+
+                if (!string.IsNullOrEmpty(reportProblem.Project))
+                {
+                    writer.WritePropertyName(ProjectProperty);
+                    writer.WriteValue(PathUtility.GetPathWithForwardSlashes(reportProblem.Project));
+                }
+
+                writer.WritePropertyName(reportProblem.ProblemType == ProblemType.Warning ? WarningProperty : reportProblem.ProblemType == ProblemType.Error ? ErrorProperty : InformationProperty);
```

Comment:
I missed this in the spec.  Doesn't this make it harder for customers to parse/filter?

In powershell, if customers only to find only errors, ignore warnings, I guess it would be something like `$errors = ConvertFrom-Json output.json | select-object { $_.problems } | where-object { $_.error -ne $null }`. Whereas, if there was a "level" property, then the last filter would be `$_.Level == "Error"`.

This also makes using a strongly typed language difficult, since reading the file needs to use polymorphic deserialization (is each object in the `Problems` array an error or warning object?). Iterating the list also then needs the `is` keyword to filter, rather than a more performant value comparison (not that I expect parsing the json output to be performance critical). for each `foreach (var problem in problems) { if (problem p is ErrorProblem e) { Console.WriteLine(e.Error); } else if (problem p is warning w) { Console.WriteLine(w.Warning); } }`. Of course, they could implement a custom json converter and their in-memory model could have properties `Level` and `Message`, even if the json file itself doesn't.  But with a slight tweak to the json schema, an auto-generated model for deserialization is possible, reducing other people's workloads.

Reply:
Sure, it's useful for end users. I added `level` property into spec.
Could you please take a look if it looks good then I can implement it.
https://github.com/NuGet/Home/blob/557f5dfdcdaf21f7a83dab6d30a35d05d33d489b/proposed/2022/DotnetListPackageMachineReadableJsonOutput.md#-dotnet-list-package---format-json-2

Reply:
`level` property is already implemented.
Please review again.